File: thumbnail/src/main/java/com/cisco/d3a/filemon/impl/ThumbnailCapturer.java
package com.cisco.d3a.filemon.impl;

import java.io.File;
import java.io.FileFilter;
import java.io.IOException;
import java.util.ArrayList;
import java.util.List;

import org.apache.commons.io.FileUtils;
import org.apache.commons.io.FilenameUtils;
import org.apache.commons.lang.StringUtils;
import org.apache.http.HttpResponse;
import org.apache.http.client.HttpClient;
import org.apache.http.client.methods.HttpDelete;
import org.apache.http.client.methods.HttpPost;
import org.apache.http.entity.ContentType;
import org.apache.http.entity.StringEntity;
import org.apache.http.util.EntityUtils;
import org.json.JSONObject;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;
import org.springframework.beans.factory.DisposableBean;
import org.springframework.beans.factory.InitializingBean;

import com.cisco.d3a.filemon.StatisticsHandler;
import com.cisco.d3a.filemon.api.ActionContext;
import com.cisco.d3a.filemon.api.CaptureSpec;
import com.cisco.d3a.filemon.api.FileProcessor;
import com.cisco.d3a.filemon.api.ImageCapturer;
import com.cisco.d3a.filemon.api.LocalStorage;
import com.cisco.d3a.filemon.impl.img.ImageCapturerManager;
import com.cisco.d3a.filemon.impl.support.ImageResizer;
import com.cisco.d3a.filemon.util.FileHelper;
import com.cisco.d3a.filemon.util.HttpClientFactory;
import com.google.common.base.Joiner;

public class ThumbnailCapturer implements FileProcessor, InitializingBean, DisposableBean {
	private static final Logger LOGGER = LoggerFactory.getLogger(ThumbnailCapturer.class);

	private ImageCapturerManager imageCapturerManager;
    private CaptureSpec captureSpec;
    private List<CaptureSpec> resizeSpecs;
    private LocalStorage localStorage;
	private String endpoint;
	private StatisticsHandler statisticsHandler;
	private String urlPrefix;
	
	public String getUrlPrefix() {
		return urlPrefix;
	}

	public void setUrlPrefix(String urlPrefix) {
		this.urlPrefix = urlPrefix;
	}

	public void setStatisticsHandler(StatisticsHandler statisticsHandler) {
		this.statisticsHandler = statisticsHandler;
	}
	
	public void setLocalStorage(LocalStorage localStorage) {
		this.localStorage = localStorage;
	}

	public void setEndpoint(String endpoint) {
		this.endpoint = endpoint;
	}

	public void setImageCapturerManager(ImageCapturerManager imageCapturerManager) {
		this.imageCapturerManager = imageCapturerManager;
	}

    public void setCaptureSpec(CaptureSpec captureSpec) {
        this.captureSpec = captureSpec;
    }

    public void setResizeSpecs(List<CaptureSpec> resizeSpecs) {
        this.resizeSpecs = resizeSpecs;
    }

    public void onModified(File file, ActionContext context) throws Exception {
		final String fullPath = context.getPath();
		
		String ext = FilenameUtils.getExtension(fullPath);

		long before = System.currentTimeMillis();
		ImageCapturer imageCapture = imageCapturerManager.getImageCapturer(ext);
		if(imageCapture == null) {
			LOGGER.debug("Unsupported file type: " + ext);
			return;
		}
		
        File imageFolder = localStorage.getTempFile(context.getPath(), true);
		
        File tImageFolder = new File(imageFolder, captureSpec.getPath());
        tImageFolder.mkdirs();
        File thumbFolder = localStorage.getThumbnailFile(context.getPath(), true);
        boolean thumbnailGenerated = false;
        try {
			LOGGER.info("Capture thumbnails for " + context.getPath());           
			imageCapture.capture(file, tImageFolder, captureSpec);
            thumbnailGenerated = true;
        } catch (IOException e) {
            LOGGER.error("Error in capturing thumbnail for " + fullPath, e);
        } catch (Exception e) {
            LOGGER.error("Error in capturing thumbnail for " + fullPath, e);
        }

		if(thumbnailGenerated) {
            List<File> thumbnailFiles = FileHelper.sort(file, tImageFolder, new FileFilter() {
    			@Override
    			public boolean accept(File pathname) {
    				return pathname.getName().endsWith(captureSpec.getType());
    			}        	
            });
            
            if(thumbnailFiles.size() != 0) {
            	statisticsHandler.incFileCaptured(1);
                resizeThumbnails(file, thumbFolder, thumbnailFiles, context);
                LOGGER.info("Thumbnail " + fullPath + " captured in " + (System.currentTimeMillis() - before) + " ms.");
            } else {
                LOGGER.info("No thumbnails captured for " + fullPath);
            }

            try {
				uploadThumbnails(file, thumbFolder, context);
			} catch(Exception e) {
				LOGGER.error(e.getMessage(), e);
			} finally {
//				FileHelper.deleteFile(imageFolder);
            }
		}
	}

    private void resizeThumbnails(File file, File to, List<File> thumbnailFiles, ActionContext context) {
        final int numberOfImages = thumbnailFiles.size();
        // Delete old thumbnail images
        for(CaptureSpec resizeSpec : resizeSpecs) {
            File dest = new File(to, resizeSpec.getPath());
            try {
				FileUtils.deleteDirectory(dest);
			} catch (IOException e) {
			}
        }        
        for(int i = 0; i < numberOfImages; i ++) {
            try {
                ImageResizer imageResizer = new ImageResizer(thumbnailFiles.get(i));
                for(CaptureSpec resizeSpec : resizeSpecs) {
                    int numOfSpecImgs = resizeSpec.getCapturePages(numberOfImages);
                    if(i < numOfSpecImgs) {
                        File dest = resizeSpec.getOutputFile(to, file, i + 1);
                        CaptureSpec.Size size = resizeSpec.getCaptureSize(imageResizer.getImage());
                        if(size.isValid()) {
                            imageResizer.resize(dest, size.width, size.height, resizeSpec.getType());
                        } else {
                        	LOGGER.warn("Illegal thumbnail size: " + size);
                        }
                    }
                }
            } catch(IOException e) {
                LOGGER.error("Error in resizing thumbnail for " + thumbnailFiles.get(i).getAbsolutePath(), e);
            }
        }
    }

    private void uploadThumbnails(File file, File thumbnailFolder, ActionContext context) throws Exception {
		if(thumbnailFolder.exists() && thumbnailFolder.isDirectory()) {
			LOGGER.debug("Uploading thumbnails under " + thumbnailFolder.getCanonicalPath());
			
			HttpClient client = HttpClientFactory.createHttpClient();
			try {
				int c = 0;
	            JSONObject thumbnails = new JSONObject();
	            for(CaptureSpec resizeSpec : resizeSpecs) {
	            	 File t = new File(thumbnailFolder, resizeSpec.getPath());
	                List<String> collector = new ArrayList<String>();
	                c += collectThumbnailsForSpec(context, t, resizeSpec, collector);
	                thumbnails.put(resizeSpec.getPath(), Joiner.on("|").join(collector.iterator()));
	            }

	            // upload thumbnails stat info
	            if(c != 0) {
		            if(c != 0) LOGGER.info(thumbnailFolder.getName() + " collected " + c + " thumbnails");

		            HttpPost post = new HttpPost(context.getServiceLocation(endpoint));
	                post.addHeader("Authorization", "Basic " + context.getToken());

	                JSONObject jsonObject = new JSONObject().put("thumbnail", thumbnails);
	                StringEntity entity = new StringEntity(jsonObject.toString(), ContentType.create("application/json", "UTF-8"));
	                post.setEntity(entity);

	                LOGGER.debug("Thumbnail req: " + jsonObject.toString());
	                HttpResponse response = client.execute(post);
	                String status = EntityUtils.toString( response.getEntity(), "UTF-8" );
	                if(response.getStatusLine().getStatusCode() == 200) {
	                    LOGGER.info("Thumbnail data for " + context.getPath() + " -> " + StringUtils.abbreviate(status, 40) + "(" + status.length() + ")");
	                } else {
	                    LOGGER.warn("Thumbnail data for " + context.getPath() + " -> " + status);
	                }
	            } else {
	            	// delete deprecated thumbnails info
					HttpDelete deleteMethod = new HttpDelete(context.getServiceLocation(endpoint));
					deleteMethod.addHeader("Authorization", "Basic " + context.getToken());
					HttpResponse response = client.execute(deleteMethod);
					LOGGER.debug(context.getPath() + " deleted thumbnails: " + EntityUtils.toString( response.getEntity(), "UTF-8" ));
					
	            }
	            statisticsHandler.incThumbnailCaptured(c);
			} finally {
				client.getConnectionManager().shutdown();		
			}
        } else {
			LOGGER.warn("Invalid thumbnail folder: " + thumbnailFolder.getName() + " for " + context.getPath());
		}
	}
	
	private int collectThumbnailsForSpec(ActionContext context, File folder, final CaptureSpec resizeSpec, List<String> collector) throws Exception {
		int count = 0;	
        List<File> thumbnails = FileHelper.sort(null, folder, new FileFilter() {
        	@Override
        	public boolean accept(File filename) {
        		return filename.getName().endsWith(resizeSpec.getType()); // same image suffix
        	}
        });
		for(int i=0;thumbnails != null && i<thumbnails.size();i++) {
			File t = thumbnails.get(i);
			StringBuilder fullPath = new StringBuilder();
			fullPath.append(this.urlPrefix);
			fullPath.append("/");
			fullPath.append(context.getPath());
			fullPath.append("/");
			fullPath.append(resizeSpec.getPath());
			fullPath.append("/");
			fullPath.append(t.getName());			
			collector.add(fullPath.toString());
			count ++;
		}
		return count;
	}

	@Override
	public void afterPropertiesSet() throws Exception {
		for(CaptureSpec resizeSpec : this.resizeSpecs) {
			LOGGER.info("ResizeSpec " + resizeSpec);
		}
		LOGGER.info("File thumbnail endpoint: " + endpoint);
		LOGGER.info("File thumbnail url prefix: " + urlPrefix);
	}
	
	@Override
	public void destroy() throws Exception {
	}
}

File: playdemo/app/helper/filemon/impl/ThumbnailCapturer.scala
package helper.filemon.impl

import helper.filemon.CaptureSpec
import java.io.File
import helper.filemon.api.ActionContext
import org.apache.commons.io.FilenameUtils
import org.joda.time.DateTime
import helper.filemon.api.ImageCapturer
import helper.filemon.MSOfficeImageCapturer
import helper.filemon.api.LocalStorage
import play.api.Logger
import java.io.IOException
import org.apache.commons.io.FileUtils
import helper.filemon.CaptureSpecSize
import helper.filemon.impl.support.ImageResizer
import org.apache.http.client.HttpClient
import scala.collection.mutable.ArrayBuffer
import play.api.libs.json._
import scala.collection.mutable.HashMap

object ThumbnailCapturer {
  val resizeSpecs = List[CaptureSpec](new CaptureSpec("large", "png", "L", 100, 0, 0, -1),
    new CaptureSpec("medium", "png", "M", 100, 480, 480, -1), new CaptureSpec("small", "png", "S", 100, 180, 180, -1))
  val urlPrefix: String = "api/file/data"
  val endPoint: String = ""
}

class ThumbnailCapturer {

  def onModified(file: File, context: ActionContext) = {
    try {
      val fullPath: String = context.path
      val ext: String = FilenameUtils.getExtension(fullPath)
      val before = DateTime.now().getMillis()
      val localStorage: LocalStorage = new LocalStorageImpl
      val captureSpec = new CaptureSpec("origin", "png", "ORIG", 100, 0, 0, -1)

      val imageCapture: ImageCapturer = new MSOfficeImageCapturer
      val imageFolder: File = localStorage.getTempFile(context.path, true);

      val tImageFolder = new File(imageFolder, captureSpec.path);
      tImageFolder.mkdirs();

      val thumbFolder: File = localStorage.getThumbnailFile(context.path, true);
      var thumbnailGenerated = false;

      try {
        Logger.info("Capture thumbnails for " + context.path);
        imageCapture.capture(file, tImageFolder, captureSpec)
        thumbnailGenerated = true
      } catch {
        case x: IOException => {
          Logger.error("Error in capturing thumbnail for " + fullPath);
          println (x)
        }
        case x: Exception => {
          Logger.error("Error in capturing thumbnail for " + fullPath);
            println (x)
        }
      }
      
      if (thumbnailGenerated) {
        println("ext:"+ captureSpec.getType)
        
        val thumbnails: List[File] =tImageFolder.listFiles().filter(file =>
        		file.getName().endsWith(captureSpec.getType)).sortWith((f1,f2)=>
           FilenameUtils.getBaseName(f1.getName()).toInt <
           FilenameUtils.getBaseName(f2.getName()).toInt
            ).toList

        if (!thumbnails.isEmpty) {
          resizeThumbnails(file, thumbFolder, thumbnails, context);
          Logger.info("Thumbnail " + fullPath + " captured in " + (System.currentTimeMillis() - before) + " ms.");
        } else {
          Logger.info("No thumbnails captured for " + fullPath);
        }
        
         try {
        	 uploadThumbnails(file, thumbFolder, context);
			} catch {
			  case x:Exception => Logger.error(x.getMessage())
			  
			}
      }

    } catch {
      case x: Exception => throw x
    }
  }

  def resizeThumbnails(file: File, to: File, thumbnailFiles: List[File], context: ActionContext) {
    val numberOfImages = thumbnailFiles.size
    // Delete old thumbnail images
    for (resizeSpec <- ThumbnailCapturer.resizeSpecs) {
      val dest = new File(to, resizeSpec.path);
      try {
        FileUtils.deleteDirectory(dest);
      } catch {
        case x: IOException => throw x
      }
    }

    for (i <- 0 until numberOfImages) {
      try {
        val imageResizer = new ImageResizer(thumbnailFiles(i))
        val timage = imageResizer.getImage
        ThumbnailCapturer.resizeSpecs.foreach { resizeSpec =>

          val numOfSpecImgs = resizeSpec.getCapturePages(numberOfImages);

          if (i < numOfSpecImgs) {
            val dest: File = resizeSpec.getOutputFile(to, file, i + 1);
            val size: CaptureSpecSize = resizeSpec.getCaptureSize(imageResizer.getImage);
            if (size.isValid) {
              imageResizer.resize(dest, size.width, size.height, resizeSpec.picType);
            } else {
              Logger.warn("Illegal thumbnail size: " + size);
            }

          }
        }
      } catch {
        case x: IOException =>
          throw x
          Logger.error("Error in resizing thumbnail for " + thumbnailFiles(i).getAbsolutePath());
      }

    }
  }
     
  	@throws[Exception]
      private def uploadThumbnails( file:File,  thumbnailFolder:File,  context:ActionContext)= {
		if(thumbnailFolder.exists() && thumbnailFolder.isDirectory()) {
			Logger.info("Uploading thumbnails under " + thumbnailFolder.getCanonicalPath());
			
//			val client:HttpClient = HttpClientFactory.createHttpClient();
			try {
				var c = 0;
//	            var thumbnails = new JSONObject();
				var thumbnails = JsObject(Seq())
	            ThumbnailCapturer.resizeSpecs.foreach {resizeSpec =>
	            	val folder= new File(thumbnailFolder, resizeSpec.path);
	                
	                val thumbArr = collectThumbnailsForSpec(context, folder, resizeSpec);
//	                thumbnails(resizeSpec.path) = thumbArr.mkString("|")
	                thumbnails = thumbnails.+((resizeSpec.path, JsString(thumbArr.mkString("|"))))
	                c += thumbArr.size
	                println(thumbArr.mkString("|"))
	                println(c)
	            }
	            println(Json.toJson(thumbnails))

			} finally {
//				client.getConnectionManager().shutdown();		
			}
        } else {
			Logger.warn("Invalid thumbnail folder: " + thumbnailFolder.getName() + " for " + context.path);
		}
	}
	
  	@throws[Exception]
	private def collectThumbnailsForSpec(context:ActionContext, folder:File, resizeSpec:CaptureSpec):ArrayBuffer[String]= {
      
  	      val thumbnails: List[File] =folder.listFiles().filter(file =>
        		file.getName().endsWith(resizeSpec.getType)).sortWith((f1,f2)=>
           FilenameUtils.getBaseName(f1.getName()).toInt <
           FilenameUtils.getBaseName(f2.getName()).toInt
            ).toList
        
        val thumbArr = new ArrayBuffer[String]()
        
		thumbnails.foreach { t=>
			 thumbArr += Array(ThumbnailCapturer.urlPrefix,context.path,resizeSpec.path,t.getName).mkString("/")
        }
		thumbArr
		
	}
  
  
  
  
}